redis: Add username

kajinamit opened the pull request:
Redis introduced ACL feature in 4.0.0, and this feature is supported by redis-py since 3.4.0[1]. When ACL is enabled, authentication requires username in addition to password.

This adds the username argument to Redis backend and Redis Sentinel backend, so that username can be passed in arguments along with password, instead of using client_kwargs and sentinel_kwargs.

[1] https://github.com/redis/redis-py/commit/8df8cd54d135380ad8b3b8807a67a3e6915b0b49

zzzeek commented:
hey there -

great , this needs a changelog file (See [example](https://github.com/sqlalchemy/dogpile.cache/commit/e3187d80896078fdd03377a374978f42e8ad28ec#diff-b1f9f4edb9991d26b08c2eee004f64e5fdefe848db5ece5686ef3b1bbab2f7bc) and also these parameters need to be annotated with:

```
.. versionadded:: 1.3.1 added 'username' parameter

```

please send a ping here as I dont get emails for new git commits on PRs, thanks!

kajinamit commented:
Hi @zzzeek 
Thanks for the prompt feedback ! I've added a changelog and pushed the new version.
I also noticed that password parameter was not documented in RedisSentinelBackend so I've also fixed in this commit.

kajinamit commented:
I noticed I have to fix one wrong class name after I submitted the comment. I've pushed the new version with it fixed.

I've named the changelog file according to the PR number but please let me know in case if we have a better naming rule.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e296dc442e77913d8b9b3ad9f5bb33262a76e91b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change e296dc442e77913d8b9b3ad9f5bb33262a76e91b: https://gerrit.sqlalchemy.org/c/sqlalchemy/dogpile.cache/+/5149

zzzeek commented on docs/build/changelog.rst:
```diff
@@ -3,7 +3,7 @@ Changelog
 =========
 
 .. changelog::
-    :version: 1.3.1
```
mmm revert this, we are still on 1.3.0.  1.3.1 is the next version.  we have automated scripts that keep this correct

zzzeek commented on docs/build/unreleased/251.rst:
```diff
@@ -0,0 +1,7 @@
+    :tags: usecase, redis
+
+    Added new paramref:`.RedisBackend.username` parameter. This is used for
+    authentication in Redis when RBAC is enabled. Aslo added new
```
typo, "aslo"

kajinamit replied:
Fixed in the latest version.

zzzeek commented on docs/build/unreleased/251.rst:
```diff
@@ -0,0 +1,7 @@
+    Added new paramref:`.RedisBackend.username` parameter. This is used for
+    authentication in Redis when RBAC is enabled. Aslo added new
+    paramref:`.RedisSentinelBackend.username` parameter, which is used for
+    ahtentication in both Redis and Sentinel when RBAC is enabled.
```
typo, "authentication" :)

kajinamit replied:
Fixed in the latest version.

zzzeek commented on dogpile/__init__.py:
```diff
@@ -1,4 +1,4 @@
-__version__ = "1.3.1"
+__version__ = "1.3.2"
```
revert this

kajinamit replied:
Fixed in the latest version.

zzzeek commented on dogpile/cache/backends/redis.py:
```diff
@@ -95,13 +97,16 @@ class RedisBackend(BytesBackend):
 
      .. versionadded:: 1.1.6  Added ``connection_kwargs`` parameter.
 
+     .. versionadded:: 1.3,2  Added ``username`` parameter.
```
write it like this:

    .. versionadded:: 1.3.1  Added ``username`` parameter.

kajinamit replied:
Fixed in the latest version.

zzzeek commented on dogpile/cache/backends/redis.py:
```diff
@@ -292,6 +302,8 @@ class RedisSentinelBackend(RedisBackend):
      asynchronous runners, as they run in a different thread than the one
      used to create the lock.
 
+     .. versionadded:: 1.3,2  Added ``username`` parameter.
```
write it like this:

    .. versionadded:: 1.3.1  Added ``username`` parameter.

kajinamit replied:
Fixed in the latest version.

zzzeek commented on tests/cache/test_redis_backend.py:
```diff
@@ -149,6 +149,7 @@ def test_connect_with_defaults(self, MockStrictRedis):
         # The defaults, used if keys are missing from the arguments dict.
         arguments = {
             "host": "localhost",
+            "username": None,
```
can we omit username here?   existing code that doesnt send a username in should continue to work

kajinamit replied:
I followed how password is tested but I agree testing the actual omit case is better. Updated to test without these optional arguments

zzzeek reviewed (changes requested):
thanks!   mostly spelling / formatting issues also the version is already bumped

kajinamit commented:
@zzzeek Thanks for reviews ! Sorry seems the spell check didn't work for some reason... I think I fixed all these problems pointed but please let me know if anything is still left.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision de0f35ba2bf3bb41defabef1c770cbdec91dd203 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset de0f35ba2bf3bb41defabef1c770cbdec91dd203 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/dogpile.cache/+/5149

kajinamit commented:
@zzzeek I've fixed the failure detected by black.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 172c195b05194cca6a982bf7e5314291fe36798c of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
**Michael Bayer** (zzzeek) wrote:

great, thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/dogpile.cache/+/5149
